realm/realm-library/src/main/cpp/io_realm_internal_OsMap.cpp

```diff
@@ -56,83 +56,146 @@ JNIEXPORT jobject JNICALL
+                        return JavaClassGlobalDef::new_object_id(env, value.get_object_id());
+                    case DataType::Type::Timestamp:
+                        return JavaClassGlobalDef::new_date(env, value.get_timestamp());
+                    case DataType::Type::Decimal:
+                        return JavaClassGlobalDef::new_decimal128(env, value.get_decimal());
+                    default:
+                        throw std::logic_error("'getValue' method only suitable for int, double, boolean, String, byte[], float, UUID, Decimal128 and ObjectId.");
+                }
+            }
+        }
     }
     CATCH_STD()
 
     return nullptr;
 }
 
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_OsMap_nativeGetMixedPtr(JNIEnv *env, jclass, jlong map_ptr,
+                                                 jstring j_key) {
+    try {
+        auto& dictionary = *reinterpret_cast<realm::object_store::Dictionary*>(map_ptr);
+        JStringAccessor key(env, j_key);
+        const Optional<Mixed>& optional_result = dictionary.try_get_any(StringData(key));
+        if (optional_result) {
+            return reinterpret_cast<jlong>(new JavaValue(from_mixed(optional_result.value())));
+        }
+    }
+    CATCH_STD();
+    return -1;
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_OsMap_nativeGetRow(JNIEnv* env, jclass, jlong map_ptr,
+                                          jstring j_key) {
+    try {
+        auto& dictionary = *reinterpret_cast<realm::object_store::Dictionary*>(map_ptr);
+        JStringAccessor key(env, j_key);
+        const Optional<Mixed>& optional_result = dictionary.try_get_any(StringData(key));
+        if (optional_result) {
+            const Mixed& value = optional_result.value();
+            if (!value.is_null()) {
+                return optional_result.value().get<ObjKey>().value;
+            }
+        }
+    }
+    CATCH_STD()
+
+    return -1;
+}
+
 JNIEXPORT void JNICALL
 Java_io_realm_internal_OsMap_nativePutMixed(JNIEnv* env, jclass, jlong map_ptr, jstring j_key,
                                             jlong mixed_ptr) {
     try {
-        auto &dictionary = *reinterpret_cast<realm::object_store::Dictionary*>(map_ptr);
-        auto &mixed = *reinterpret_cast<Mixed*>(mixed_ptr);
+        auto& dictionary = *reinterpret_cast<realm::object_store::Dictionary*>(map_ptr);
+        auto mixed_java_value = *reinterpret_cast<JavaValue*>(mixed_ptr);
+        const Mixed& mixed = mixed_java_value.to_mixed();
         JStringAccessor key(env, j_key);
+        dictionary.insert(StringData(key), mixed);
+    }
+    CATCH_STD()
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_OsMap_nativePutNull(JNIEnv* env, jclass, jlong map_ptr,
+                                           jstring j_key) {
+    try {
+        auto& dictionary = *reinterpret_cast<realm::object_store::Dictionary*>(map_ptr);
+        JStringAccessor key(env, j_key);
+        dictionary.insert(StringData(key).data(), Mixed());
```

Comment:
Using `Mixed()` instead of `Any()` to handle null values simplifies things a lot. Thanks to Nikola for the suggestion.